msal/application.py

```diff
@@ -304,7 +304,30 @@ def __init__(
 
             .. admonition:: Supporting raw assertion obtained from elsewhere
 
+                *Added in version 1.32.0*:
+
+                MSAL Python accepts a callback which shall obtain an assertion::
+
+                    import msal
+
+                    def assertion_callback():
+                        return '''A fresh JWT with claims:
+                            aud, exp, iss, jti, nbf, and sub'''
+
+                    smart_callback = msal.AutoRefresher(
+                        assertion_callback,
+                        expires_in=3600,  # Whatever value that you see fit
```

Comment:
Not sure what this auto refresher is. We should mention FIC scenarios in the docs instead.

Reply:
> Not sure what this auto refresher is.

Naming is hard, and new name suggestions are welcome (perhaps `CallbackCache` instead?). But the functionality behind the "auto refresher" was needed since day 1, because even when in the legit authenticated-by-a-cert scenario, MSAL Python internally has (1) a helper function to use that cert to create a *short-lived* assertion, and (2) a helper class `AutoRefresher` to cache the callback outcome and only call the callback again after a predefined expiry time elapsed. Such a cache behavior is desirable for the `#2` scenario described in [this comment](https://github.com/AzureAD/microsoft-authentication-library-for-python/issues/746#issuecomment-2359581836).

In this PR, we repurpose the `#1` callback behavior to accept any callback; and expose the `#2` cache behavior so that app developers don't need to reinvent the wheel.

> We should mention FIC scenarios in the docs instead.

I suppose, the `#1` callback can be used in any scenario including FIC, but the `#2` cache concept shall always be applicable.

Reply:
FIC example would be good. You can use example from here: https://review.learn.microsoft.com/en-us/identity/microsoft-identity-platform/federated-identity-credentials?branch=main